### ISSUE TITLE
[Clang] Assertion failure on structured binding declared with an explicit array type

### ISSUE THREAD

**Issue by akashagrwl** [llvm/llvm-project#210960]
Clang (assertions-enabled build, trunk) crashes on an assertion when a structured binding / decomposition declaration is given an explicit (non-auto) array type instead of auto.

Godbolt repro- https://godbolt.org/z/c4fjeaPxf

```
typedef int array_type[3];
void f() {
  int arr[4]{1, 2, 3, 4};
  array_type [a, b, c]{arr};
}
```
Expected: a diagnostic that a structured binding's declared type must be auto or a reference to auto (the declaration is ill-formed).

Actual:Assertion `S.Context.hasSameUnqualifiedType(SubInit[0]->getType(), DestType) && "Deduced to other type?"' failed.
